sys/include/clif.h

```diff
@@ -0,0 +1,316 @@
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Return types for the @ref sys_clif API
+ */
+enum {
+    CLIF_OK          = 0,    /**< success */
+    CLIF_NO_SPACE    = -1,   /**< not enough space in the buffer */
+    CLIF_NOT_FOUND   = -2    /**< could not find a component in a buffer */
+};
+
+/**
+ * @brief Types of link format attributes
+ */
+typedef enum {
+    CLIF_ATTR_ANCHOR        = 0,  /**< anchor */
+    CLIF_ATTR_REL           = 1,  /**< rel */
+    CLIF_ATTR_LANG          = 2,  /**< hreflang */
+    CLIF_ATTR_MEDIA         = 3,  /**< media */
+    CLIF_ATTR_TITLE         = 4,  /**< title */
+    CLIF_ATTR_TITLE_EXT     = 5,  /**< title* */
+    CLIF_ATTR_TYPE          = 6,  /**< type */
+    CLIF_ATTR_RT            = 7,  /**< rt */
+    CLIF_ATTR_IF            = 8,  /**< if */
+    CLIF_ATTR_SZ            = 9,  /**< sz */
+    CLIF_ATTR_CT            = 10, /**< ct */
+    CLIF_ATTR_OBS           = 11, /**< obs */
+    CLIF_ATTR_EXT           = 12  /**< extensions */
+} clif_attr_type_t;
+
+/**
+ * @brief Link format attribute descriptor
+ */
+typedef struct {
+    char *value;                   /**< string with the value */
+    unsigned value_len;            /**< length of the value */
+    const char *key;               /**< attribute name */
+    unsigned key_len;              /**< length of the attribute name */
+} clif_attr_t;
+
+/**
+ * @brief Link format descriptor
+ */
+typedef struct {
+    char *target;                 /**< target string */
+    unsigned target_len;          /**< length of target string */
+    clif_attr_t *attrs;         /**< array of attributes */
+    unsigned attrs_len;          /**< size of array of attributes */
+} clif_t;
+
+/**
+ * @brief Encodes a given link in link format into a given buffer
+ *
+ * @pre `link != NULL`
+ *
+ * @param[in] link      link to encode.Must not be NULL.
+ * @param[out] buf      buffer to output the encoded link. Can be NULL
+ * @param[in] maxlen    size of @p buf
+ *
+ * @note If @p buf is NULL this will return the amount of bytes that would be
+ *       needed
+ *
+ * @return amount of bytes used from @p buf in success
+ * @return CLIF_NO_SPACE if there is not enough space in the buffer
+ */
+ssize_t clif_encode_link(const clif_t *link, char *buf, size_t maxlen);
+
+/**
+ * @brief   Decodes a string of link format. It decodes the first occurrence of
+ *          a link.
+ *
+ * @pre `(link != NULL) && (buf != NULL)`
+ *
+ * @param[out] link         link to populate. Must not be NULL.
+ * @param[in]  attrs        array of attrs to populate
+ * @param[in]  attrs_len    length of @p attrs
+ * @param[in]  buf          string to decode. Must not be NULL.
+ * @param[in]  maxlen       size of @p buf
+ *
+ * @return number of bytes parsed from @p buf in success
+ * @return CLIF_NOT_FOUND if the string is malformed
+ */
+ssize_t clif_decode_link(clif_t *link, clif_attr_t *attrs, unsigned attrs_len,
+                         char *buf, size_t maxlen);
```

Comment:
`buf` should be a const char *. I believe there are a few other instances like this. Please review the rest of the API.